### ISSUE TITLE
New Antlers Engine - recursive children

### ISSUE THREAD

**Issue by danijelk** [statamic/cms#6319]
### Bug description

Whilst building our new nav with a depth of 3, we saw that using recursive method would be ideal.

But quickly did we see that the depth variable gets reset on the recursive loop, is this the intended behaviour? If it is it seems impossible to know which depth one is really in?

<img width="965" alt="image" src="https://user-images.githubusercontent.com/580753/178115620-44f673c2-f4b8-4c0e-8633-04a4b6c0b904.png">

Outputting {{ title }} - {{ depth }}

<img width="316" alt="image" src="https://user-images.githubusercontent.com/580753/178115651-62c21cfe-ab32-4571-84ca-05fa4d2b22d5.png">

### How to reproduce

Create a nav with a depth of 3.

### Versions

Statamic 3.3.18 Pro
Laravel 9.19.0
PHP 8.1.8

### Installation

Starter Kit using via CLI

### Antlers Parser

runtime (new)

**Comment by JohnathonKoster** [comment]
Out of curiosity do you have more than one nav in that template? I was able to reproduce this when there are multiple navs (but this is a bug), and am testing a fix for that now. Just want to be sure that the setup is the same 🙂

**Comment by danijelk** [comment]
Hiya @JohnathonKoster 
We do have multiple navs, I tried to disable them but the result is the same. 

After now spending a few minutes actually digging into the RecursiveNodeManager, I found what I was looking for, as I thought, as soon as we use recirsive keyword it resets the depth to 0 as the refId changes of the node

```
    public static function incrementDepth(AntlersNode $node)
    {
        if (! array_key_exists($node->refId, self::$depthMapping)) {
            self::$depthMapping[$node->refId] = 0;
        }
```

But luckily we have a secondary (undocumented?) variable we can use called 

```
        $namedDepthMapping = $node->content.'_depth';

        if (! array_key_exists($namedDepthMapping, self::$namedDepthMapping)) {
            self::$namedDepthMapping[$namedDepthMapping] = 0;
        }
```

which solves our recursive depth issue, and we get the depth we actually were looking for :)

{{ title }} - {{ depth }} - {{ children_depth }}

![image](https://user-images.githubusercontent.com/580753/178124246-1f28f962-cb05-4381-aa9f-0b1dbeb450fb.png)

**Comment by JohnathonKoster** [comment]
Ah interesting. Would you be able to share the original template that caused the issue? May be able to get something worked out for that 👍

**Comment by danijelk** [comment]
Yes for sure, https://gist.github.com/danijelk/d9db41fb04b063a6fffed23c05010919

Although I'm not seeing a way to accomplish what I intended (mainly not setting grid on navs that only have a maxdepth 2, and grids on children that have children maxdepth od 3 or above :) The recursive wizardry is indeed powerful, but I feel like I might be missing a if children:has_children kind of method, that I would normally do in blade/php.

**Comment by JohnathonKoster** [comment]
Thanks for that! I think I see where it's getting confused, and may have a solution for it but need to think on it some more.

Thanks again!!
